src/unraphael/dash/pages/5_cluster.py

```diff
@@ -0,0 +1,298 @@
+from __future__ import annotations
+import numpy as np
+import streamlit as st
+#from skimage import color
+from unraphael.dash.image_clustering import (align_images_to_mean, equalize_images,
+                                             cluster_images, compute_metrics,
+                                             clustimage_clustering, plot_clusters,
+                                             plot_dendrogram, matrix_of_similarities)
+from styling import set_custom_css
+from widgets import load_images_widget, show_heatmaps_widget
+#import matplotlib.pyplot as plt
+#import seaborn as sns
+#import io
+
+def show_images_widget(
+    images: dict[str, np.ndarray],
+    *,
+    n_cols: int = 4,
+    key: str = 'show_images',
+    message: str = 'Select image',
+) -> None | str:
+    """Widget to show images with given number of columns."""
+    col1, col2 = st.columns(2)
+    
+    # Ensure each number_input has a unique key
+    n_cols_key = f'{key}_cols'
+    n_cols = col1.number_input(
+        'Number of columns for display', value=4, min_value=1, step=1, key=n_cols_key
+    )    
+
+    cols = st.columns(n_cols)
+
+    for i, (name, im) in enumerate(images.items()):
+        if i % n_cols == 0:
+            cols = st.columns(n_cols)
+        col = cols[i % n_cols]
+
+        col.image(im, use_column_width=True, caption=name,clamp=True)    
+
+
+def equalize_images_widget(*, images: dict[str, np.ndarray]) -> dict[str, np.ndarray]:
```

Comment:
Refactor this widget and combine it with the one from the compare page.

Reply:
I refactored this widget. 

I'm not sure, though, if combining them would make things easier:
- In the compare page, the variables such as brightness are set equal to the value from the template image. Informing on these metrics would then not provide really interesting information
- In the cluster page, these parameters are set to the mean values for all images (if selected). These are interesting to show, since it informs us on the degree of equalization
- the compare page would not benefit from adding the option to equalize based on the 'mean image', since it's functionality is centered on visually comparing with a template/baseline image
- plotting the mean aligned image as an extra option in the compare page would also not be helpful, as it usually will be a bit blurry

- in the compare page, the alignment takes place because we want to visually inspect the differences compared to a template/baseline image
- in the cluster page, the alignment takes place as a necesarry preprocessing step for the cluster analysis

What would be the best approach to tackle this?

Reply:
Oh, I see, one equalizes to the base_image and the other equalizes to the average. Then this comment does not make much sense, I was confused by that they have the same function name.

Might be worth to rename the widgets then, e.g.:

- equalize_images -> equalize_images_to_base